File: ops/lib/run-cmd.js
/**
 * Subprocess wrapper for the ops layer. Every ops command that shells
 * out (launchctl, tar, sudo, curl-via-fetch helpers, the apple-docs
 * binary itself) goes through this single entry point so:
 *
 *  - timeouts are uniform and explicit (`deadlineMs`)
 *  - stdout + stderr capture is bounded (no OOM from a wedged child)
 *  - non-zero exits and timeouts throw the same structured error
 *  - tests inject a `spawn` fake instead of really running the command
 *
 * Distinct from `src/lib/spawn-with-deadline.js` (which we keep) because
 * the ops use-case wants tee-to-log behaviour and a slightly different
 * return shape — and we'd rather not couple ops back into `src/`.
 */

const DEFAULT_DEADLINE_MS = 60_000
const DEFAULT_STDERR_MAX_BYTES = 256 * 1024
const DEFAULT_STDOUT_MAX_BYTES = 4 * 1024 * 1024

/**
 * @typedef {Object} RunCmdOptions
 * @property {number} [deadlineMs=60000]
 * @property {string} [cwd]
 * @property {Record<string, string>} [env]
 * @property {'pipe' | 'inherit' | 'ignore'} [stdout='pipe']
 * @property {'pipe' | 'inherit' | 'ignore'} [stderr='pipe']
 * @property {string} [stdin]                              fed to child stdin
 * @property {number} [stdoutMaxBytes]
 * @property {number} [stderrMaxBytes]
 * @property {{ spawn?: typeof Bun.spawn, clock?: () => number }} [deps]
 *
 * @typedef {Object} RunCmdResult
 * @property {string} stdout
 * @property {string} stderr
 * @property {number} exitCode
 * @property {number} elapsedMs
 */

/**
 * Run a command to completion, throwing on non-zero exit or deadline.
 *
 * @param {string[]} args             [cmd, ...args]
 * @param {RunCmdOptions} [options]
 * @returns {Promise<RunCmdResult>}
 */
export async function runCmd(args, options = {}) {
  if (!Array.isArray(args) || args.length === 0) {
    throw new RunCmdError('runCmd: args must be a non-empty array', { args, kind: 'usage' })
  }
  const {
    deadlineMs = DEFAULT_DEADLINE_MS,
    cwd,
    env,
    stdout = 'pipe',
    stderr = 'pipe',
    stdin,
    stdoutMaxBytes = DEFAULT_STDOUT_MAX_BYTES,
    stderrMaxBytes = DEFAULT_STDERR_MAX_BYTES,
    deps = {},
  } = options
  const spawn = deps.spawn ?? Bun.spawn
  const clock = deps.clock ?? Date.now

  const startedAt = clock()
  const proc = spawn(args, {
    stdout,
    stderr,
    cwd,
    env,
    stdin,
  })

  let timedOut = false
  const timer = setTimeout(() => {
    timedOut = true
    try { proc.kill('SIGKILL') } catch { /* already dead */ }
  }, deadlineMs)

  const stdoutPromise = stdout === 'pipe' && proc.stdout
    ? readStreamCapped(proc.stdout, stdoutMaxBytes)
    : Promise.resolve('')
  const stderrPromise = stderr === 'pipe' && proc.stderr
    ? readStreamCapped(proc.stderr, stderrMaxBytes)
    : Promise.resolve('')

  const [stdoutText, stderrText, exitCode] = await Promise.all([
    stdoutPromise,
    stderrPromise,
    proc.exited,
  ])
  clearTimeout(timer)

  const elapsedMs = clock() - startedAt

  if (timedOut) {
    throw new RunCmdError(
      `runCmd: ${args[0]} exceeded ${deadlineMs}ms`,
      { args, kind: 'timeout', deadlineMs, stderr: stderrText, elapsedMs },
    )
  }

  if (exitCode !== 0) {
    throw new RunCmdError(
      `runCmd: ${args[0]} exited ${exitCode}: ${stderrText.trim().slice(0, 512) || '<no stderr>'}`,
      { args, kind: 'exit', exitCode, stderr: stderrText, stdout: stdoutText, elapsedMs },
    )
  }

  return { stdout: stdoutText, stderr: stderrText, exitCode, elapsedMs }
}

/**
 * Same as runCmd, but does NOT throw on non-zero exit. Use when the
 * caller wants to branch on the exit code (e.g. `launchctl print` to
 * test whether a label is loaded — exit 113 means "not present").
 *
 * @param {string[]} args
 * @param {RunCmdOptions} [options]
 * @returns {Promise<RunCmdResult>}
 */
export async function runCmdAllowFailure(args, options = {}) {
  try {
    return await runCmd(args, options)
  } catch (err) {
    if (err instanceof RunCmdError && err.kind === 'exit') {
      return {
        stdout: err.stdout ?? '',
        stderr: err.stderr ?? '',
        exitCode: err.exitCode,
        elapsedMs: err.elapsedMs ?? 0,
      }
    }
    throw err
  }
}

export class RunCmdError extends Error {
  constructor(message, { args, kind, exitCode, deadlineMs, stdout, stderr, elapsedMs } = {}) {
    super(message)
    this.name = 'RunCmdError'
    this.args = args
    this.kind = kind
    this.exitCode = exitCode
    this.deadlineMs = deadlineMs
    this.stdout = stdout
    this.stderr = stderr
    this.elapsedMs = elapsedMs
  }
}

async function readStreamCapped(stream, maxBytes) {
  const reader = stream.getReader()
  const decoder = new TextDecoder()
  let received = 0
  let truncated = false
  let out = ''
  while (true) {
    const { value, done } = await reader.read()
    if (done) break
    if (received + value.byteLength > maxBytes) {
      const room = maxBytes - received
      if (room > 0) out += decoder.decode(value.subarray(0, room), { stream: true })
      truncated = true
      try { reader.cancel() } catch {}
      break
    }
    received += value.byteLength
    out += decoder.decode(value, { stream: true })
  }
  out += decoder.decode()
  if (truncated) out += `\n…[truncated at ${maxBytes} bytes]`
  return out
}

File: swift/Sources/ADOps/RunCmd.swift
// Subprocess wrapper for the ops layer — the native port of ops/lib/run-cmd.js.
//
// Every ops verb that shells out (launchctl, install, git, caddy, the bun
// entrypoint) goes through one seam so timeouts are uniform + explicit
// (`deadlineMs`), stdout/stderr capture is bounded, and non-zero exits +
// timeouts throw the same structured error. `CommandRunner` is the injection
// point: verbs take `any CommandRunner`, tests pass a fake, production passes
// `ProcessCommandRunner`.

private import Foundation

/// How a child's stdout/stderr is handled.
public enum StdioMode: Sendable {
    case pipe
    case inherit
    case ignore
}

/// Options for a single command invocation.
public struct RunCmdOptions: Sendable {
    public var deadlineMs: Int
    public var cwd: String?
    public var env: [String: String]?
    public var stdout: StdioMode
    public var stderr: StdioMode
    public var stdin: String?
    public var stdoutMaxBytes: Int
    public var stderrMaxBytes: Int

    public init(
        deadlineMs: Int = 60_000,
        cwd: String? = nil,
        env: [String: String]? = nil,
        stdout: StdioMode = .pipe,
        stderr: StdioMode = .pipe,
        stdin: String? = nil,
        stdoutMaxBytes: Int = 4 * 1024 * 1024,
        stderrMaxBytes: Int = 256 * 1024
    ) {
        self.deadlineMs = deadlineMs
        self.cwd = cwd
        self.env = env
        self.stdout = stdout
        self.stderr = stderr
        self.stdin = stdin
        self.stdoutMaxBytes = stdoutMaxBytes
        self.stderrMaxBytes = stderrMaxBytes
    }
}

/// The result of a completed command.
public struct RunCmdResult: Sendable, Equatable {
    public let stdout: String
    public let stderr: String
    public let exitCode: Int32
    public let elapsedMs: Int

    public init(stdout: String, stderr: String, exitCode: Int32, elapsedMs: Int) {
        self.stdout = stdout
        self.stderr = stderr
        self.exitCode = exitCode
        self.elapsedMs = elapsedMs
    }
}

/// A structured subprocess failure. `kind` lets `runAllowFailure` return an exit
/// failure as a value while still rethrowing timeouts / usage errors.
public struct RunCmdError: Error, Sendable {
    public enum Kind: Sendable, Equatable { case usage, timeout, exit, spawn }
    public let message: String
    public let kind: Kind
    public let args: [String]
    public let exitCode: Int32
    public let stdout: String
    public let stderr: String
    public let elapsedMs: Int

    public init(
        message: String, kind: Kind, args: [String] = [], exitCode: Int32 = 0,
        stdout: String = "", stderr: String = "", elapsedMs: Int = 0
    ) {
        self.message = message
        self.kind = kind
        self.args = args
        self.exitCode = exitCode
        self.stdout = stdout
        self.stderr = stderr
        self.elapsedMs = elapsedMs
    }
}

/// The subprocess seam. `run` throws on non-zero exit / timeout; `runAllowFailure`
/// returns a non-zero exit as a value (rethrows timeout / usage / spawn).
public protocol CommandRunner: Sendable {
    func run(_ args: [String], options: RunCmdOptions) async throws -> RunCmdResult
    func runAllowFailure(_ args: [String], options: RunCmdOptions) async throws -> RunCmdResult
}

extension CommandRunner {
    public func run(_ args: [String]) async throws -> RunCmdResult {
        try await run(args, options: RunCmdOptions())
    }
    /// Default `runAllowFailure`: call `run`, convert an `exit` failure to a value.
    public func runAllowFailure(_ args: [String], options: RunCmdOptions) async throws
        -> RunCmdResult
    {
        do {
            return try await run(args, options: options)
        } catch let error as RunCmdError where error.kind == .exit {
            return RunCmdResult(
                stdout: error.stdout, stderr: error.stderr, exitCode: error.exitCode,
                elapsedMs: error.elapsedMs)
        }
    }
}

/// The production runner: `Foundation.Process` with a SIGKILL deadline.
public struct ProcessCommandRunner: CommandRunner {
    private let nowMs: @Sendable () -> Double

    public init(nowMs: @escaping @Sendable () -> Double = ProcessCommandRunner.systemNowMs) {
        self.nowMs = nowMs
    }

    /// Wall clock in epoch-millis (body uses Foundation; the type does not).
    public static let systemNowMs: @Sendable () -> Double = {
        Date().timeIntervalSince1970 * 1000
    }

    public func run(_ args: [String], options: RunCmdOptions) async throws -> RunCmdResult {
        guard let executable = args.first else {
            throw RunCmdError(message: "runCmd: args must be a non-empty array", kind: .usage)
        }
        let started = nowMs()
        let process = Process()
        process.executableURL = URL(fileURLWithPath: executable)
        process.arguments = Array(args.dropFirst())
        if let cwd = options.cwd { process.currentDirectoryURL = URL(fileURLWithPath: cwd) }
        if let env = options.env { process.environment = env }

        let outPipe = configureStdout(process, options.stdout)
        let errPipe = configureStderr(process, options.stderr)
        configureStdin(process, options.stdin)

        do {
            try process.run()
        } catch {
            throw RunCmdError(
                message: "runCmd: cannot spawn \(executable): \(error)", kind: .spawn, args: args)
        }

        async let outText = readCapped(outPipe, maxBytes: options.stdoutMaxBytes)
        async let errText = readCapped(errPipe, maxBytes: options.stderrMaxBytes)
        let timedOut = await raceExitOrDeadline(process, deadlineMs: options.deadlineMs)
        let stdoutText = await outText
        let stderrText = await errText
        let elapsed = Int(nowMs() - started)

        if timedOut {
            throw RunCmdError(
                message: "runCmd: \(executable) exceeded \(options.deadlineMs)ms",
                kind: .timeout, args: args, stderr: stderrText, elapsedMs: elapsed)
        }
        let code = process.terminationStatus
        if code != 0 {
            let snippet = stderrText.trimmingCharacters(in: .whitespacesAndNewlines)
            let brief = snippet.isEmpty ? "<no stderr>" : String(snippet.prefix(512))
            throw RunCmdError(
                message: "runCmd: \(executable) exited \(code): \(brief)", kind: .exit, args: args,
                exitCode: code, stdout: stdoutText, stderr: stderrText, elapsedMs: elapsed)
        }
        return RunCmdResult(
            stdout: stdoutText, stderr: stderrText, exitCode: code, elapsedMs: elapsed)
    }
}

// MARK: - Process plumbing

private func configureStdout(_ process: Process, _ mode: StdioMode) -> Pipe? {
    switch mode {
        case .pipe:
            let pipe = Pipe()
            process.standardOutput = pipe
            return pipe
        case .ignore:
            process.standardOutput = FileHandle.nullDevice
            return nil
        case .inherit:
            return nil
    }
}

private func configureStderr(_ process: Process, _ mode: StdioMode) -> Pipe? {
    switch mode {
        case .pipe:
            let pipe = Pipe()
            process.standardError = pipe
            return pipe
        case .ignore:
            process.standardError = FileHandle.nullDevice
            return nil
        case .inherit:
            return nil
    }
}

private func configureStdin(_ process: Process, _ stdin: String?) {
    guard let stdin else { return }
    let pipe = Pipe()
    process.standardInput = pipe
    let handle = pipe.fileHandleForWriting
    handle.write(Data(stdin.utf8))
    try? handle.close()
}

/// Read a pipe to EOF (or `maxBytes`), off the cooperative pool.
private func readCapped(_ pipe: Pipe?, maxBytes: Int) async -> String {
    guard let pipe else { return "" }
    return await withCheckedContinuation { (continuation: CheckedContinuation<String, Never>) in
        DispatchQueue.global()
            .async {
                let handle = pipe.fileHandleForReading
                var collected = Data()
                while collected.count < maxBytes {
                    let chunk = handle.availableData
                    if chunk.isEmpty { break }
                    collected.append(chunk)
                }
                let capped = collected.count > maxBytes ? collected.prefix(maxBytes) : collected[...]
                continuation.resume(returning: String(decoding: capped, as: UTF8.self))
            }
    }
}

/// Resolve to `true` iff the deadline fired first (and the child was SIGKILLed).
/// The continuation is claimed exactly once by whichever path wins.
private func raceExitOrDeadline(_ process: Process, deadlineMs: Int) async -> Bool {
    let claim = ResumeOnce()
    return await withCheckedContinuation { (continuation: CheckedContinuation<Bool, Never>) in
        DispatchQueue.global()
            .async {
                process.waitUntilExit()
                if claim.tryClaim() { continuation.resume(returning: false) }
            }
        DispatchQueue.global()
            .asyncAfter(deadline: .now() + .milliseconds(max(0, deadlineMs))) {
                if claim.tryClaim() {
                    kill(process.processIdentifier, SIGKILL)
                    continuation.resume(returning: true)
                }
            }
    }
}

/// A one-shot claim guard, so the shared continuation resumes exactly once.
private final class ResumeOnce: @unchecked Sendable {
    private let lock = NSLock()
    private var claimed = false
    func tryClaim() -> Bool {
        lock.lock()
        defer { lock.unlock() }
        if claimed { return false }
        claimed = true
        return true
    }
}
